### PR TITLE
Add support for running hooks.

### PR DESCRIPTION
**Pull request by kyrofa**
This involves a move away from "appname" to "security tag".

### CODE REVIEW

File: src/sc-main.c [hunk 1]
```diff
@@ -54,19 +54,19 @@ int sc_main(int argc, char **argv)
 	if (argc < NR_ARGS + 1)
 		die("Usage: %s <security-tag> <binary>", argv[0]);
 
-	const char *appname = argv[1];
-	debug("appname is %s", appname);
+	const char *executable_name = argv[1];
```
Comment (zyga) [inline src/sc-main.c]: In reality this is `secuirty_tag`.

Comment (zyga) [inline src/sc-main.c]: If you are changing this, can you just call it as such, and remove `aa_profile` entirely as it was always the same thing

File: src/snap.c [hunk 2]
```diff
@@ -23,19 +23,22 @@
...
+	// snap.<name>.(<appname>|hook.<hookname>)
 	// - <name> must start with lowercase letter, then may contain
 	//   lowercase alphanumerics and '-'
-	// - <app> may contain alphanumerics and '-'
-	const char *whitelist_re = "^snap\\.[a-z][a-z0-9-]*\\.[a-zA-Z0-9-]+$";
+	// - <appname> may contain alphanumerics and '-'
+	// - <hookname must start with a lowercase letter, then may
+	//   contain lowercase letters and '-'
+	const char *whitelist_re =
+	    "^snap\\.[a-z](-?[a-z0-9])*\\.([a-zA-Z0-9](-?[a-zA-Z0-9])*|hook\\.[a-z](-?[a-z])*)$";
```
Comment (zyga) [inline src/snap.c:36]: Can you perhaps link to the snapd code that does similar checks?

File: src/verify-executable-name-test.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+#include "snap.h"
+#include "snap.c"
+
+#include <glib.h>
+
+static void test_verify_executable_name()
+{
+	// First, test the names we know are good
```
Comment (zyga) [inline src/verify-executable-name-test.c:25]: Nice, thank you!

File: src/verify-executable-name-test.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	g_assert_true(verify_executable_name("snap.name.app"));
+	g_assert_true(verify_executable_name
+		      ("snap.network-manager.NetworkManager"));
+	g_assert_true(verify_executable_name("snap.f00.bar-baz1"));
+	g_assert_true(verify_executable_name("snap.foo.hook.bar"));
+	g_assert_true(verify_executable_name("snap.foo.hook.bar-baz"));
+
```
Comment (zyga) [inline src/verify-executable-name-test.c:32]: Can you add a canary test for `"snap.name"`. At some point we may decide to abbreviate the security tag and this would be a nice test to catch that change.

File: src/snap.c [hunk 5]
```diff
@@ -23,19 +23,22 @@
...
-	// snap.<name>.<app>
+	// The executable name is of form:
+	// snap.<name>.(<appname>|hook.<hookname>)
 	// - <name> must start with lowercase letter, then may contain
 	//   lowercase alphanumerics and '-'
-	// - <app> may contain alphanumerics and '-'
-	const char *whitelist_re = "^snap\\.[a-z][a-z0-9-]*\\.[a-zA-Z0-9-]+$";
+	// - <appname> may contain alphanumerics and '-'
+	// - <hookname must start with a lowercase letter, then may
+	//   contain lowercase letters and '-'
```
Comment (zyga) [inline src/snap.c:34]: can hook names contain digits?

File: src/snap.c [hunk 6]
```diff
@@ -23,19 +23,22 @@
...
+	const char *whitelist_re =
+	    "^snap\\.[a-z](-?[a-z0-9])*\\.([a-zA-Z0-9](-?[a-zA-Z0-9])*|hook\\.[a-z](-?[a-z])*)$";
 	regex_t re;
 	if (regcomp(&re, whitelist_re, REG_EXTENDED | REG_NOSUB) != 0)
 		die("can not compile regex %s", whitelist_re);
 
-	int status = regexec(&re, appname, 0, NULL, 0);
```
Comment (zyga) [inline src/snap.c:38]: Just thinking about something nasty being done. Can we require that total length of `security_tag` be smaller than some sane default. I don't want to let people explode the system by installing a snap with a long app name that is expensive to check.

If there is something in snapd, let's use that limit, if not then let's make up one now.

@jdstrand ^^ FYI

File: src/snap.c (hunk 5 above)
Comment (kyrofa) [inline src/snap.c:34]: No, just lowercase letters and hyphens.

File: src/snap.c (hunk 2 above)
Comment (kyrofa) [inline src/snap.c:36]: It's [here](https://github.com/snapcore/snapd/blob/master/snap/validate.go), are you saying you want that link in the code?

File: src/sc-main.c (hunk 1 above)
Comment (kyrofa) [inline src/sc-main.c]: Yeah might as well.

File: src/snap.c (hunk 6 above)
Comment (kyrofa) [inline src/snap.c:38]: I don't know of a limit imposed by snapd, so throw out a number! Though that may be something for another PR.

Comment (jdstrand) [inline src/snap.c:38]: There are practical limits to how long this can be and the review tools enforce this. The max length enforced by the store is '100' for the full security label ('snap.$snap.$app').

File: src/sc-main.c (hunk 1 above)
Comment (kyrofa) [inline src/sc-main.c]: Done.

File: src/verify-executable-name-test.c (hunk 4 above)
Comment (kyrofa) [inline src/verify-executable-name-test.c:32]: Done.

File: src/snap.c (hunk 6 above)
Comment (kyrofa) [inline src/snap.c:38]: @zyga do you want that in this PR, or should it be done separately?

File: src/snap.c (hunk 5 above)
Comment (zyga) [inline src/snap.c:34]: OK, thanks

File: src/snap.c (hunk 6 above)
Comment (zyga) [inline src/snap.c:38]: This can go in as there wasn't any check around there before. I can take care of that.

**Comment by zyga** [comment]
+1, thanks for working on this
